src/beast/beast/asio/handler_alloc.h

```diff
@@ -0,0 +1,306 @@
+            p, size, h_);
+    }
+
+#ifdef _MSC_VER
+    // Work-around for MSVC not using allocator_traits
+    // in the implementation of shared_ptr
+    //
+    void
+    destroy(T* t)
+    {
+        t->~T();
+    }
+#endif
+
+    template<class U>
+    friend
+    bool
+    operator==(handler_alloc const& lhs,
+        handler_alloc<U, Handler> const& rhs)
+    {
+        return true;
+    }
+
+    template<class U>
+    friend
+    bool
+    operator!=(handler_alloc const& lhs,
+        handler_alloc<U, Handler> const& rhs)
+    {
+        return !(lhs == rhs);
+    }
+};
+
+template <class T, class Handler>
+class no_handler_alloc :
+    public std::allocator<T>
+{
+private:
+    using alloc_type = std::allocator<T>;
+
+    // We want a partial template specialization as a friend
+    // but that isn't allowed so we friend all versions. This
+    // should produce a compile error if Handler is not
+    // constructible from H.
+    //
+    template <class U, class H>
+    friend class no_handler_alloc;
+
+public:
+    using value_type = T;
+    using pointer = value_type*;
+    using reference = value_type&;
+    using const_pointer = value_type const*;
+    using const_reference = value_type const&;
+    using difference_type = std::ptrdiff_t;
+    using size_type = std::size_t;
+
+    template <class U>
+    struct rebind
+    {
+    public:
+        using other = no_handler_alloc<U, Handler>;
+    };
+
+    no_handler_alloc() = delete;
+    no_handler_alloc(no_handler_alloc&&) = default;
+    no_handler_alloc(no_handler_alloc const&) = default;
+    no_handler_alloc& operator=(no_handler_alloc&&) = default;
+    no_handler_alloc& operator=(no_handler_alloc const&) = default;
+
+    no_handler_alloc(Handler&&)
+    {
+    }
+
+    no_handler_alloc(Handler const&)
+    {
+    }
+
+    template<class U>
+    no_handler_alloc(
+        no_handler_alloc<U, Handler>&& other)
+    {
+    }
+
+    template<class U>
+    no_handler_alloc(
+        no_handler_alloc<U, Handler> const& other)
+    {
+    }
+
+#ifdef _MSC_VER
+    // Work-around for MSVC not using allocator_traits
+    // in the implementation of shared_ptr
+    //
+    void
+    destroy(T* t)
+    {
+        t->~T();
+    }
+#endif
+
+    template<class U>
+    friend
+    bool
+    operator==(no_handler_alloc const& lhs,
+        no_handler_alloc<U, Handler> const& rhs)
+    {
+        return true;
+    }
+
+    template<class U>
+    friend
+    bool
+    operator!=(no_handler_alloc const& lhs,
+        no_handler_alloc<U, Handler> const& rhs)
+    {
+        return !(lhs == rhs);
+    }
+};
+
+//------------------------------------------------------------------------------
+
+template<class Handler>
+class temp_buffer
+{
+    Handler& h_;
+    std::size_t n_ = 0;
+    std::uint8_t* p_ = nullptr;
+
+public:
+    explicit
+    temp_buffer(Handler& h)
+        : h_(h)
+    {
+    }
+
+    ~temp_buffer()
+    {
+        if(p_)
+            dealloc();
+    }
```

Comment:
Because of the ownership of `p_`, it looks like `temp_buffer` needs a user-declared copy constructor (possibly deleted).  If the copy constructor is declared/deleted, and a move constructor is desired, that needs to be user-provided.  Both assignment operators are implicitly deleted as the class stands now (which I presume is fine).

A deleted copy constructor compiles.